contrib/node/src/python/pants/contrib/node/tasks/node_resolve.py

```diff
@@ -5,15 +5,58 @@
 from __future__ import absolute_import, division, print_function, unicode_literals
 
 import os
+from hashlib import sha1
 
+from future.utils import PY3
 from pants.base.build_environment import get_buildroot
+from pants.base.fingerprint_strategy import DefaultFingerprintHashingMixin, FingerprintStrategy
 from pants.base.workunit import WorkUnitLabel
 from pants.build_graph.build_graph import sort_targets
 
 from pants.contrib.node.tasks.node_paths import NodePaths, NodePathsLocal
 from pants.contrib.node.tasks.node_task import NodeTask
 
 
+class NodeResolveFingerprintStrategy(DefaultFingerprintHashingMixin, FingerprintStrategy):
+  """
+  Fingerprint package lockfiles (e.g. package.json, yarn.lock...),
+  so that we don't automatically run this if none of those have changed.
+
+  We read every file and add its contents to the hash.
+  """
+
+  def __init__(self, node_paths):
+    """
+    :param node_paths NodePathsBase:
+        Collection of resolved paths, which may be relative to the working dir.
+        Particularly, it offers per-target querying of those paths.
+        Used to guess the location of the _files_to_watch
+    """
+    self.node_paths = node_paths
+
+  _files_to_watch = [
+    "package.json",
+    "yarn.lock"
+  ]
+
+  def _get_path_base(self, target):
+    if self.node_paths.node_path(target) is not None:
+      return self.node_paths.node_path(target)
+    else:
+      return target.address.spec_path
```

Comment:
The recommend approach is that these files are always included in the target's sources: see https://www.pantsbuild.org/node_readme.html#package-management ... and I believe that pants renders a warning if you don't add the lockfile to `sources`.

So I think that it would be less surprising for this to just always fingerprint those files from the target's sources, and drop usage of the `node_paths`.

Reply:
You read my mind, I noted that as a TODO while working on the CLI option :)

Reply:
Done